### ISSUE TITLE
Support DynamicProxy configuration to not write back "ref" args

### ISSUE THREAD

**Issue by bruthig-g** [castleproject/Core#449]
Migrating from [StackOverflow](https://stackoverflow.com/questions/56977070/can-castle-dynamicproxy-be-configured-to-not-write-to-ref-args) as per jonorossi's suggestion.

The offending third-party interface is [IDebugProperty2](https://docs.microsoft.com/en-us/visualstudio/extensibility/debugger/reference/idebugproperty2-enumchildren?view=vs-2015) provided by Microsoft.  And more specifically it's the "ref Guid guidFilter" arg of the EnumChildren() method that fails to be assigned to.

I have confirmed that my implementation of EnumChildren (see below) that is **not** wrapped in a proxy will also throw an AccessViolationException.

```
public int EnumChildren (enum_DEBUGPROP_INFO_FLAGS   dwFields, uint dwRadix, ref Guid                    guidFilter, uint dwAttribFilter, string pszNameFilter, uint dwTimeout, out IEnumDebugPropertyInfo2 ppEnum)
{
  guidFilter = Guid.NewGuid();
} 
```

<StackOverflowDescription>

I'm hitting an AccessViolationException when trying to use an interface proxy that is wrapping a third-party interface. The exception is being thrown when the emitted code writes the invocation arg array back to the original arg for a 'ref' type arg. Presumably this is because the third-party component has given an 'in' only parameter and it should not be written to.

So, I was wondering if DynamicProxy could be configured (or enhanced/hacked) to not write back some args? Maybe by configuring a custom code emitter or something...

Or thinking outside of the box, would it work to apply an [InAttribute] at runtime to the third-party interface (assuming DynamicProxy can be configured to adhere to this attribute)?

Other, maybe useful, information:

The third-party interface doesn't use the 'in' keyword or InAttribute on the offending 'ref' arg.
I'm trying to use interface proxies created with a target
Example code:

```
namespace third_party
    {
      public class IService
      {
        public int DoSomething(ref Guid guid);
      }
    }
    namespace domain_ns
    {
      public class ServiceImpl : IService
      {
        public int DoSomething(ref Guid guid) { return 0; }
      }

      public static void main(string[] args)
      {
        ...
        var proxy = proxyGen.CreateInterfaceProxyWithTarget(..., new ServiceImpl(), ... );
        third_party.AddService(proxy);

        // Indirectly causes proxy.DoSomething() to be called.
        // Throws an AccessViolationException :(
        third_party.Go();
      }
    }
```

</StackOverflowDescription>

**Comment by stakx** [comment]
Hi @bruthig-g, could you please provide an MCVE (minimal, complete, and verifiable code example) that we can run to observe the `AccessViolationException`? Otherwise, it's hard to do anything, as we don't know what exactly is going wrong.

> I was wondering if DynamicProxy could be configured (or enhanced/hacked) to not write back some args? Maybe by configuring a custom code emitter or something...

You can create your own fork of DynamicProxy and edit its source code. (I haven't checked in detail, but I believe [this line of code](https://github.com/castleproject/Core/blob/f0626e2cb2a8df9da113bed8abcbfc7d50631d63/src/Castle.Core/DynamicProxy/Generators/InvocationTypeGenerator.cs#L171) is the approximate location where you'd want to make changes.)

Speaking about the main project here, in my opinion, adding a configurable "write lock" feature for C# `ref` parameters to DynamicProxy doesn't make much sense: These parameters explicitly allow writes. If DynamicProxy didn't write back the argument to `ref` parameters, changes made in interceptors wouldn't propagate back to the caller, which as far as I can see would be considered a bug in the great majority of all use cases... so we'd have a niche feature that most people would never want to use.

In your Stack Overflow question, you wrote:

>  Presumably this is because the third-party component has given an 'in' only parameter and it should not be written to.

If I understand you correctly, somehow they manage to do something like this:

```csharp
int Foo(IService service, in Guid guid) => service.DoSomething(ref guid);
```

If so, then obviously that's the problem to be fixed.

**Comment by stakx** [comment]
P.S. regarding my above statement:

> I believe [this line of code](https://github.com/castleproject/Core/blob/f0626e2cb2a8df9da113bed8abcbfc7d50631d63/src/Castle.Core/DynamicProxy/Generators/InvocationTypeGenerator.cs#L171) or [this one](https://github.com/castleproject/Core/blob/f0626e2cb2a8df9da113bed8abcbfc7d50631d63/src/Castle.Core/DynamicProxy/Generators/MethodWithInvocationGenerator.cs#L130) seem plausible locations [...]

At first glance, I suspect that we might have a bug at the first code location that I linked to: Unlike in the second location, there appears to be no check that distinguishes between `ref` and `in` parameters... they're apparently all written back. This probably deserves a closer look.

**Comment by bruthig-g** [comment]
> You can create your own fork of DynamicProxy and edit its source code. (I haven't checked in detail, but I believe [this line of code](https://github.com/castleproject/Core/blob/f0626e2cb2a8df9da113bed8abcbfc7d50631d63/src/Castle.Core/DynamicProxy/Generators/InvocationTypeGenerator.cs#L171) or [this one](https://github.com/castleproject/Core/blob/f0626e2cb2a8df9da113bed8abcbfc7d50631d63/src/Castle.Core/DynamicProxy/Generators/MethodWithInvocationGenerator.cs#L130) seem plausible locations where you might make changes.)

Thanks, this is very helpful!

> Speaking about the main project here, in my opinion, adding a configurable "write lock" feature for C# `ref` parameters to DynamicProxy doesn't make much sense: These parameters explicitly allow writes. If DynamicProxy didn't write back the argument to `ref` parameters, changes made in interceptors wouldn't propagate back to the caller, which as far as I can see would be considered a bug in the great majority of all use cases... so we'd have a niche feature that most people would never want to use.

Unfortunately I'm inclined to agree with you :(

> In your Stack Overflow question, you wrote:
> 
> > Presumably this is because the third-party component has given an 'in' only parameter and it should not be written to.
> 
> If I understand you correctly, somehow they manage to do something like this:
> 
> ```cs
> int Foo(IService service, in Guid guid) => service.DoSomething(ref guid);
> ```

Yea, I suspect they've done something silly like this.  Unfortunately getting this fixed by the third-party (i.e. Microsoft) doesn't really seem to be an option.  Especially since end users would have to upgrade Visual Studio in order to use our product.


Another idea has occurred to me.  It appears DP is checking against System.ReflectionParameterAttributes.In, I wonder if I'd be able to add this to the arg at runtime before DP processes it.  Would it be possible to somehow update the ParameterInfo.Attributes before DP processes it and decides whether to write back to the arg?

**Comment by stakx** [comment]
> Would it be possible to somehow update the ParameterInfo.Attributes before DP processes it and decides whether to write back to the arg?

I don't see how this would be easily possible.

It seems you have a choice between hacking the CLR (runtime), hacking DynamicProxy, or perhaps, in this instance, take the easy way out and write a proxy type manually in order to get full control.

There's still a small chance here that DynamicProxy is at fault somehow, so if you have a repro for us to look at, that'd be great.

**Comment by stakx** [comment]
> I suspect that we might have a bug at the first code location that I linked to [...].

I took that closer look, everything is in order: `in` parameters (like all others) are buffered in the `invocation.Arguments` array at the start of interception, and are then treated as `ref` arguments during interception. That is, an interceptor can pass a different value to the subsequent interceptor or target. (The first code location that I linked to above concerns itself with passing a value to the method on the proxy target.) However, `in` parameters are *not* written back to the caller at the end of interception. (The second code location that I linked to above concerns itself with that part.)

**Comment by jonorossi** [comment]
> I don't think DP is doing anything wrong here. It should be copying "ref" args back to the caller. The problem is the third_party behaves poorly and uses a "ref" arg to pass write-protected memory and they rely on "documentation" to make sure people don't assign to it.
> (from StackOverflow)

@bruthig-g I didn't know the AccessViolationException was being thrown by the third party library and not DP when you logged this. That definitely makes it much more difficult and not something we can really fix as @stakx has definitely explained well.

If I were you I'd probably just abstract away from the `IDebugProperty2` interface, it looks like an ugly interface anyway so just add another layer of indirection which would also probably mean you don't need a proxy. Hacking DP, VS or the CLR are all pretty terrible solutions in my opinion.

**Comment by bruthig-g** [comment]
Hey,

So after thinking about this a little more I'm wondering if you'd be willing to adopt the following solution in to Castle.Core that only writes back the ref arg if the identity has actually changed.  I'd argue that this isn't offensively hacky and would make Castle.Core more resilient to misbehaving third-party libraries.

Here's an example of how the the emitted code would look like for the 
`IDebugProperty2.EnumChildren(enum_DEBUGPROP_INFO_FLAGS dwFields, uint dwRadix, ref Guid guidFilter, uint dwAttribFilter, string pszNameFilter, uint dwTimeout, out IEnumDebugPropertyInfo2 ppEnum)` method.  The additions are lines 13,14,16.

```
1.     public override void InvokeMethodOnTarget()
2.     {
3.       this.EnsureValidTarget();
4.       Guid argumentValue = (Guid) this.GetArgumentValue(2);
5.       IEnumDebugPropertyInfo2 ppEnum = (IEnumDebugPropertyInfo2) this.GetArgumentValue(6);
6.       int num;
7.       try
8.       {
9.         num = (this.target as IDebugProperty2).EnumChildren((enum_DEBUGPROP_INFO_FLAGS) this.GetArgumentValue(0), (uint) this.GetArgumentValue(1), ref argumentValue, (enum_DBG_ATTRIB_FLAGS) this.GetArgumentValue(3), (string) this.GetArgumentValue(4), (uint) this.GetArgumentValue(5), out ppEnum);
10.      }
11.      finally
12.      {
13.        if (!Object.ReferenceEquals(Arguments[2], argumentValue))
14.        {
15.          this.SetArgumentValue(2, (object) argumentValue);
16.        }
17.        this.SetArgumentValue(6, (object) ppEnum);
18.      }
19.      this.ReturnValue = (object) num;
20.    }
```

What do you think?

Thanks,
Ben

**Comment by stakx** [comment]
@bruthig-g, let me start by saying that I think it's great that you're taking the time to take a closer look at DP and see how it works at a lower level, then come back with a concrete proposal. :+1:

> What do you think?

**TL;DR:** To be honest, I am not thrilled by the idea of spending implementation time & adding complexity to DP just to better support what I assume is faulty external code. The cost is counteracted by comparatively little benefit. But I probably wouldn't object if someone else wanted to do all that work.

Here's the longer version:

1. By-refs generally need to be written back at the "boundaries" of the interceptor part of the interception pipeline. There are two such boundaries where this needs to happen:

   a. When interception returns from the target object to the last interceptor
   b. When interception returns from the first interceptor to the calling code.

   (See also https://github.com/castleproject/Core/issues/449#issuecomment-511016849 above.)

   The additional check would most likely only be necessary for (b), but not for (a). This is because (AFAIK) inside the interception pipeline, all by-ref arguments essentially count as `ref` ones, even those that were originally `in` arguments. You're therefore not going to prevent any `AccessViolationException` with an additional check in (b).

   I believe your code example is looking at (b) when it really should be looking at (a). But let's ignore that detail for the moment, your code example is still sufficiently accurate to keep the discussion going.

2. Your code example contains a small bug, because boxing and unboxing of value-typed arguments will change object identity. In line 4, you're extracting the third argument and cast it to `Guid`. This causes an unboxing operation. In line 13, you're passing that `Guid` to `object.ReferenceEquals(object, object)`, which causes another boxing operation. Because of that, the check will always yield `false`, even when the target didn't write to that particular argument. The following changes might be necessary:

   ```diff
    1.     public override void InvokeMethodOnTarget()
    2.     {
    3.       this.EnsureValidTarget();
   -4.       Guid argumentValue = (Guid) this.GetArgumentValue(2);
   +         object argumentValueBoxed = this.GetArgumentValue(2);
   +4.       Guid argumentValue = (Guid) argumentValueBoxed;
    5.       IEnumDebugPropertyInfo2 ppEnum = (IEnumDebugPropertyInfo2) this.GetArgumentValue(6);
    6.       int num;
    7.       try
    8.       {
    9.         num = (this.target as IDebugProperty2).EnumChildren((enum_DEBUGPROP_INFO_FLAGS) this.GetArgumentValue(0), (uint) this.GetArgumentValue(1), ref argumentValue, (enum_DBG_ATTRIB_FLAGS) this.GetArgumentValue(3), (string) this.GetArgumentValue(4), (uint) this.GetArgumentValue(5), out ppEnum);
    10.      }
    11.      finally
    12.      {
   -13.        if (!Object.ReferenceEquals(Arguments[2], argumentValue))
   +13.        if (!object.ReferenceEquals(this.GetArgumentValue(2), argumentValueBoxed))
    14.        {
    15.          this.SetArgumentValue(2, (object) argumentValue);
    16.        }
    17.        this.SetArgumentValue(6, (object) ppEnum);
    18.      }
    19.      this.ReturnValue = (object) num;
    20.    }
   ```

   What this means is mostly more complexity during implementation: You'll need to only add the additional check for `ref` arguments, but not for `out` ones. You'll need one local variable for reference-typed `ref` arguments, but *two* for value-typed ones (or, if you favor ease of implementation over runtime efficiency, just track all `ref` arguments with two locals).

3. This extra check for all `ref` arguments introduces an additional overhead, both at proxy generation time (more code needs to be generated), and possibly at runtime. I freely admit that this additional overhead is most probably negligible and may even counteract the cost of unnecessary write-backs at runtime.

   Nonetheless, it's something extra we have to do, and take the time to implement... and all just to support misbehaving external code. That last bit is why I personally am not much thrilled by your proposal.

I would go along with what @jonorossi proposed above: Create a thin abstraction (anti-corruption layer) around that Microsoft interface so you have something cleaner to work with. In the best case, you will no longer need a proxy; in the second-best case, this additional abstraction will become your proxy's target... in both cases, no changes would be required in DP.

**Comment by bruthig-g** [comment]
Thanks for the detailed info @stakx !

I agree I only care about case (b) when the first interceptor returns to the calling code.  I guess this means I should be looking to patch the GeneratorUtil.CopyOutAndrefParameter() method to produce the following generated code instead:

```diff
  1.     [MethodImpl(MethodImplOptions.PreserveSig)]
  2.     public int IDebugProperty3\u002EEnumChildren(
  3.       [ComAliasName("Microsoft.VisualStudio.Debugger.Interop.DEBUGPROP_INFO_FLAGS"), In] enum_DEBUGPROP_INFO_FLAGS dwFields,
  4.       [In] uint dwRadix,
  5.       [In] ref Guid guidFilter,
  6.       [ComAliasName("Microsoft.VisualStudio.Debugger.Interop.DBG_ATTRIB_FLAGS"), In] enum_DBG_ATTRIB_FLAGS dwAttribFilter,
  7.       [In] string pszNameFilter,
  8.       [In] uint dwTimeout,
  9.       out IEnumDebugPropertyInfo2 ppEnum)
  10.    {
  11.      IDebugProperty3_EnumChildren property3EnumChildren = new IDebugProperty3_EnumChildren((IDebugProperty3) null, (object) this, this.__interceptors, IDebugProperty3Proxy.token_EnumChildren_1, new object[7]
  12.      {
  13.        (object) dwFields,
  14.        (object) dwRadix,
  15.        (object) guidFilter,
  16.        (object) dwAttribFilter,
  17.        (object) pszNameFilter,
  18.        (object) dwTimeout,
  19.        (object) ppEnum
  20.      });
  21.      try
  22.      {
  23.        ((AbstractInvocation) property3EnumChildren).Proceed();
  24.      }
  25.      finally
  26.      {
  27.        object[] arguments = ((AbstractInvocation) property3EnumChildren).get_Arguments();
+ 28.        if (guidFilter.Equals(arguments[2]))
+ 29.        {
  30.            guidFilter = (Guid) arguments[2];
+ 31         }
  32.        ppEnum = (IEnumDebugPropertyInfo2) arguments[6];
  33.      }
  34.      return (int) ((AbstractInvocation) property3EnumChildren).get_ReturnValue();
  35.    }
```

Of course I would have to use `object.ReferenceEquals()` for any non-value type args instead of `.Equals()`.

I've taken a stab at a patch but I'm running in to problems decompiling the code so I can't figure out why it doesn't work.  Any chance you could point me in the right direction based on the following work in progress?

```diff
 1.   public static void CopyOutAndRefParameters(TypeReference[] dereferencedArguments, LocalReference invocation,
 2.                                              MethodInfo method, MethodEmitter emitter)
 3.   {
 4.       var parameters = method.GetParameters();
 5.
 6.       // Create it only if there are byref writable arguments.
 7.       LocalReference arguments = null;
 8.
 9.       for (var i = 0; i < parameters.Length; i++)
 10.      {
 11.          if (IsByRef(parameters[i]) && !IsReadOnly(parameters[i]))
 12.          {
 13.              if (arguments == null)
 14.              {
 15.                  arguments = StoreInvocationArgumentsInLocal(emitter, invocation);
 16.              }
+17.              var elementType = parameters[i].ParameterType.GetElementType();
+18.              if (elementType.IsValueType)
+19.              {
+20.                  var equalsMethod = elementType.GetMethod("Equals", new[] { typeof(object) });
+21.                  var equalsCall = new MethodInvocationExpression(dereferencedArguments[i], equalsMethod, new LoadRefArrayElementExpression(i, arguments));
+22.                  var conditionalAssign = new IfNullExpression(equalsInvocation, AssignArgument(dereferencedArguments, i, arguments));
+23.                  emitter.CodeBuilder.AddStatement(new ExpressionStatement(conditionalAssign));
+24.              }
+25.              else
+26.              {
+27.                  // TODO: Use object.ReferenceEquals()
 28.                  emitter.CodeBuilder.AddStatement(AssignArgument(dereferencedArguments, i, arguments));
+29.              }
 30.          }
 31.      }
 32.  }
```

**Comment by stakx** [comment]
> [...] but I'm running in to problems decompiling the code so I can't figure out why it doesn't work.

I'd much rather help you with this problem. The general procedure is:

1. Create some proxies, then tell DynamicProxy to save the generated dynamic assembly to disk:

   ```csharp
   var generator = new ProxyGenerator(new PersistentProxyBuilder());
   ... // generate proxy objects here...
   generator.ProxyBuilder.ModuleScope.SaveAssembly();
   ```

2. Then open the saved assembly file (it should be called `CastleDynProxy2.dll`) with a tool like ILDASM or ILSpy and inspect the generated proxy types.

**Comment by stakx** [comment]
P.S. you can probably just use the static `object.Equals(a, b)` method for both value and reference types, as it should call `object.ReferenceEquals` as the first check for reference types.

**Comment by stakx** [comment]
@bruthig-g, this issue has been dormant for several months. Is there still anything you need, or should we close this issue?

**Comment by bruthig-g** [comment]
We don't plan to pursue any work here.  We ended up "just adding a layer of indirection" for the offending methods.  So in the example below `IDebugProperty3.EnumChildren()` is not decorated but `IMyDebugProperty.EnumChildrenImpl()` is.

I still sand by my thoughts that CastleProxy could be more resilient to misbehaving third-party code but I understand that the cost/benefit ratio might not be worth it based on this example alone.  Feel free to close.

```c#

public interface IMyDebugProperty : IDebugProperty3 {

    int EnumChildrenImpl(enum_DEBUGPROP_INFO_FLAGS fields, uint radix, Guid guidFilter,
        enum_DBG_ATTRIB_FLAGS attributeFilter, string nameFilter, uint timeout,
        out IEnumDebugPropertyInfo2 propertyEnum);
}

public abstract class MyDebugProperty : IMyDebugProperty {
        public int EnumChildren(enum_DEBUGPROP_INFO_FLAGS fields, uint radix, ref Guid guidFilter,
            enum_DBG_ATTRIB_FLAGS attributeFilter, string nameFilter,
            uint timeout, out IEnumDebugPropertyInfo2 propertyEnum)
        {
            // Converts |guidFilter| to a non ref type arg so that Castle DynamicProxy doesn't fail
            // to assign values back to it.
            return EnumChildrenImpl(fields, radix, guidFilter, attributeFilter, nameFilter,
                                         timeout, out propertyEnum);
        }
}

```

**Comment by stakx** [comment]
Thanks for replying @bruthig-g. Yes, you are certainly right that DynamicProxy *could* be somewhat more resilient against misbehaving user code, but we're likely going to focus on other things in the near future, so I am going to close this issue for now.
